File: frontend/src/types/commandCentre.ts
/** Command centre types — mirror of `app/schemas/command_centre.py`. */

export type NextAction =
  | 'respond_to_recruiter' | 'complete_assessment' | 'upload_cv' | 'upload_document'
  | 'relogin' | 'complete_verification' | 'confirm_application' | 'review_answer'
  | 'follow_up' | 'schedule_interview' | 'prepare_interview' | 'review_application'
  | 'verify_submission' | 'resume_application' | 'update_profile' | 'wait' | 'none';

export type ActionPriority = 'high' | 'medium' | 'low' | 'none';

export type ApplicationHealth =
  | 'healthy' | 'needs_attention' | 'blocked' | 'stale'
  | 'session_required' | 'document_required' | 'action_required' | 'status_unknown';

export interface ActionItem {
  application_id: string;
  job_id: string;
  title: string;
  company: string;
  action: NextAction;
  priority: ActionPriority;
  /** Plain-English justification, rendered verbatim so the ranking can be sanity-checked. */
  reason: string;
  score: number;
  due_at: string | null;
  health: ApplicationHealth;
  status: string;
  /** Which surface resolves this action, so a click lands on the exact thing to do. */
  target: string;
  application_url: string | null;
  portal: string | null;
}

export interface ActionQueue {
  items: ActionItem[];
  total: number;
  by_priority: Record<string, number>;
}

export interface CommandCentreSummary {
  total_applications: number;
  needs_attention: number;
  high_priority: number;
  by_status: Record<string, number>;
  /** Operational health, NOT the hiring status — an application can be at interview and
   *  simultaneously blocked on an expired session. */
  by_health: Record<string, number>;
  upcoming_interviews: number;
  pending_assessments: number;
  top_actions: ActionItem[];
}

export interface TimelineEntry {
  id: string;
  event_type: string;
  occurred_at: string;
  summary: string;
  detail: string | null;
  actor: string;
  payload: Record<string, unknown> | null;
}

export interface ApplicationTimeline {
  application_id: string;
  entries: TimelineEntry[];
  total: number;
  current_action: NextAction;
  current_reason: string;
}

/** Label + colour per action. Colours are semantic tokens, never literals. */
export const ACTION_META: Record<NextAction, { label: string; cta: string }> = {
  complete_assessment: { label: 'Assessment due', cta: 'Open assessment' },
  prepare_interview: { label: 'Interview upcoming', cta: 'Prepare' },
  relogin: { label: 'Session expired', cta: 'Reconnect & resume' },
  complete_verification: { label: 'Verification required', cta: 'Complete verification' },
  resume_application: { label: 'Application paused', cta: 'Resume' },
  upload_cv: { label: 'CV required', cta: 'Choose CV' },
  upload_document: { label: 'Document required', cta: 'Upload' },
  review_application: { label: 'Awaiting your review', cta: 'Review' },
  review_answer: { label: 'Answer needs review', cta: 'Review answer' },
  confirm_application: { label: 'Confirmation required', cta: 'Confirm' },
  verify_submission: { label: 'Submission unconfirmed', cta: 'Verify' },
  follow_up: { label: 'Follow-up recommended', cta: 'Follow up' },
  respond_to_recruiter: { label: 'Recruiter waiting', cta: 'Reply' },
  schedule_interview: { label: 'Interview to schedule', cta: 'Schedule' },
  update_profile: { label: 'Profile out of date', cta: 'Update' },
  wait: { label: 'In progress', cta: 'View' },
  none: { label: 'No action', cta: 'View' },
};

export const PRIORITY_COLOR: Record<ActionPriority, string> = {
  high: 'var(--rejected)',
  medium: 'var(--review)',
  low: 'var(--accent)',
  none: 'var(--text-4)',
};

File: backend/app/schemas/command_centre.py
"""Schemas for the command centre: action queue, summary, timeline."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import (
    ActionPriority,
    ApplicationEventType,
    ApplicationHealth,
    ApplicationStatus,
    NextAction,
)


class ActionItem(BaseModel):
    """One row of the action queue."""

    model_config = ConfigDict(from_attributes=True)

    application_id: str
    job_id: str
    title: str
    company: str
    action: NextAction
    priority: ActionPriority
    #: Plain-English justification, shown verbatim — the user must be able to sanity-check
    #: the ranking, or they stop trusting the queue.
    reason: str
    score: float
    due_at: datetime | None = None
    health: ApplicationHealth
    status: ApplicationStatus
    #: Which surface resolves this action, so a click lands on the exact thing to do
    #: rather than a page the user must navigate onward from.
    target: str
    application_url: str | None = None
    portal: str | None = None


class ActionQueueResponse(BaseModel):
    items: list[ActionItem] = Field(default_factory=list)
    total: int = 0
    by_priority: dict[str, int] = Field(default_factory=dict)


class CommandCentreSummary(BaseModel):
    """Everything the dashboard's top panels need, in one call."""

    total_applications: int = 0
    needs_attention: int = 0
    high_priority: int = 0
    by_status: dict[str, int] = Field(default_factory=dict)
    #: Operational health, which is not the hiring status — an application can be at
    #: INTERVIEW and simultaneously BLOCKED on an expired session.
    by_health: dict[str, int] = Field(default_factory=dict)
    upcoming_interviews: int = 0
    pending_assessments: int = 0
    top_actions: list[ActionItem] = Field(default_factory=list)


class TimelineEntry(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    event_type: ApplicationEventType
    occurred_at: datetime
    summary: str
    detail: str | None = None
    actor: str
    payload: dict | None = None


class ApplicationTimelineResponse(BaseModel):
    application_id: str
    entries: list[TimelineEntry] = Field(default_factory=list)
    total: int = 0
    current_action: NextAction = NextAction.NONE
    current_reason: str = ""
